doc + extmarks minor rename [ci skip]

justinmk opened the pull request:
- Doc: `nvim_buf_get_extmarks`: give examples of `(row, col)` tuples for range params
- rename `set_extmark_index_from_obj`
- `nvim_buf_get_extmarks`: rename "amount" to "limit"

norcalli commented:
Can you add a note that extmarks are 0-indexed. It's not exactly clear from the docs.

Also, you can't use a non-existent mark id to begin a search using mark id instead of position, i.e. it doesn't work like a btree for that part. I ran into this when I was using it, I assumed that it was using it like an upper and lower bound, but it's actually a search starting from a mark id.

bfredl commented:
> Also, you can't use a non-existent mark id to begin a search using mark id instead of position, i.e. it doesn't work like a btree for that part.

Why is this expected? You can start a search from a concrete position or a mark which actually exists in the tree. Is it expected that btree's keep a living memory of objects that have been deleted?

norcalli commented:
@bfredl it's probably because the docs use (or used) the words `lower` and `upper` to describe the bounds. That made it so I was imagining a binary tree like

```
     1
    / \
   /   \
  3     \
 / \     4
5   6   /
       9
```

In which case you could find the interval `[2, inf)` (which would be `get_extmarks(..., 2, -1)`) even though it doesn't exist.

bfredl commented:
"infinity" is possible though. `get_extmarks` accept 0 and -1 pseudo-indexes for start and end of buffer. I think this is clearly enough documented, no?

norcalli commented:
@bfredl yeah I didn't have a problem with that part. I was using math interval notation, so I wasn't going to say `[2, -1)` because that makes no sense (or is an empty interval) :P The other part was the main comment.

bfredl commented:
But what is the problem? mark 2 doesn't exist, so by definition it does not have a position. It seems like you expect some kind of secondary structure that sorts marks by id, and we don't have that (nor anything in the docs that indicate so?). Ids are just id:s, they don't "mean" anything to the API.

norcalli commented:
@bfredl I explained very clearly that I **imagined** something else, which turned out to be a false assumption. All I'm saying is that by using the words `lower` and `upper`, it implanted in my mind the notion that it could be a range query for ids. That's it. I think the the problem is that for positions, you *can* specify a position that doesn't correlate to a mark, which made me think you could do the same for marks. And I get that ids are just ids, but ids are also numbers, which are things which, in some places, can be range queried.

Why is it so difficult to add a disclaimer? Or are you looking for something else?

bfredl commented:
> I think the the problem is that for positions, you can specify a position that doesn't correlate to a mark, which made me think you could do the same for marks.

This was essentially what I just said. You expect it have a separate behaviour which makes it sort by id instead of position, and we don't have that. We only provide one ordering, which is by position only.

> Why is it so difficult to add a disclaimer? Or are you looking for something else?

maybe just "range" needs to be replaced by "buffer range" or "position range" ? or is a sentence/paragraph necessary to explain it sorts by position only?

norcalli commented:
@bfredl This whole time we've been agreeing, but I was just describing a problem I had so that an note could be made in case someone else had the same problem. I wasn't suggesting anything change in extmarks, which is what I think you misinterpreted. This is the docs PR, afterall.

Personally, I think having different behaviour in a single function is kind of meh, which is why if we go with that approach, the docs should be as verbose as possible.

I would also add valid examples:

```
local pos = nvim_win_get_cursor(0)
get_extmarks({pos[1]-1, pos[2]}, -1)
get_extmarks({pos[1]-1, pos[2]}, 0)
get_extmarks(0, -1)

get_extmarks(mark_id, -1)
```

Maybe something like:

````
NOTE: if specifying a mark id as a range bound, the mark must exist.
```

bfredl commented:
> I  wasn't suggesting anything change in extmarks, which is what I think you misinterpreted. 

I clearly didn't say anything like that either. I'm just saying that the _root cause_ of the misunderstanding is just the question of sorting order. Once it is clear that it is sorting by position and nothing else, the question of "is it required the mark exist or not" doesn't even make sense. I e we cannot "just change" to allow non-exist marks, because that was never the root cause.

norcalli commented:
I know that, but saying that it's sorted by position order might confuse people, whereas saying the mark exists implies that if you happen to delete a mark in between using its id for a query, then you could clearly understand why that is. That's basically the only time someone would likely do that.

If you want, you could add `the mark must exist. This is because extmarks are sorted by positions in the buffer.` Feel free to recommend a better solution. We've spent too many comments on this already.

bfredl commented:
I guess we have to agree to disagree then :) Sorry if my comments seemed to be personally directed to you, what I meant to say is that _anyone_ using this function need to know that operates on position ordering fundamentally. If this is not clear, we obviously need to fix the documentation. I was just curious to dig in your thinking (when you first read the docs) to see how to improve it for someone else with similar expectations to you.

justinmk commented:
see latest push. 

```
nvim_buf_get_extmarks({buffer}, {ns_id}, {start}, {end}, {opts})
                Gets extmarks in a range (inclusive).

                Range bounds can be given as (row, col) tuples, or valid
                extmark ids (whose positions define the range). 0 and -1 work
                as shorthand for (0,0) and (-1,-1) respectively, thus the
                following are equivalent:

                  all_marks = nvim_buf_get_extmarks(0, my_ns, 0, -1, {})
                  all_marks = nvim_buf_get_extmarks(0, my_ns, [0,0], [-1,-1], {})


                If end is less than start , traversal works backwards.
                (Useful with limit , to get the first marks prior to a given
                position.)

                Parameters:
                    {buffer}  Buffer handle, or 0 for current buffer
                    {ns_id}   Namespace id from nvim_create_namespace()
                    {start}   Start of range, given as (row, col) or valid
                              extmark id (whose position defines the bound)
                    {end}     End of range, given as (row, col) or valid
                              extmark id (whose position defines the bound)
                    {opts}    Optional parameters. Keys:
                              • limit: Maximum number of marks to return
```

> I would also add valid examples:

todo

justinmk committed:
doc + extmarks tweaks [ci skip]

- nvim_buf_get_extmarks: rename "amount" => "limit"
- rename `set_extmark_index_from_obj`

justinmk commented on runtime/doc/api.txt:
```diff
@@ -1797,81 +1798,93 @@ nvim_buf_get_extmark_by_id({buffer}, {ns_id}, {id})
+                    {end}     End of range, given as (row, col) or valid
+                              extmark id (whose position defines the bound)
+                    {opts}    Optional parameters. Keys:
+                              • limit: Maximum number of marks to return
```
@norcalli JFI in case you are using the `amount` param: renamed it to `limit`.

norcalli replied:
@justinmk thanks :)
